Validation

briprowe opened the pull request:
Changed how we represent config. From:

``` clojure
(defconfig
  :env {foo "FOO_VAR"
        bar "BAR_VAR"}
  :app {baz [:some :key :path]})
```

To:

``` clojure
(defconfig
  :env [[foo "FOO_VAR"]
        [bar "BAR_VAR"]]
  :app [[baz [:some :key :path]]])
```

This enables us to specify defaults and validators:

``` clojure
(defconfig
  :env [[foo "FOO_VAR" {:default "foo"}]
        [bar "BAR_VAR" {:validator schema/URL}]]
  :app [[baz [:some :key :path]]])
```

colinrymer committed:
Create .circle.yml

Commit:
Set the RP signing key as the default signing key.

Commit:
Default to using pwd as the root dir when reading app config

Commit:
Nothing ever throws an ExceptionInfo.

Since there's no code that throws an ExceptionInfo in the codebase, I
think that our tests are wrong. It's improperly enforcing the wrong
contract so to speak.

The contract appears to be that we throw FileNotFoundException when the
app-config doesn't exist. So, I've changed the test to reflect that.

Commit:
Move get-in* and ->vec to config-entry ns

Commit:
Implemtn ConfigEntry type

Commit:
Bump major version to 1.0.0

Changed how we represent config. From:

```clojure
(defconfig
  :env {foo "FOO_VAR"
        bar "BAR_VAR"}
  :app {baz [:some :key :path]})
```

To:

```clojure
(defconfig
  :env [[foo "FOO_VAR"]
        [bar "BAR_VAR"]]
  :app [[baz [:some :key :path]]])
```

whalesalad commented:
:heart_eyes:

briprowe commented:
Oh I should rebase to get that WIP out of there.

mayankag commented on src/clj_config/core.clj:
```diff
@@ -183,4 +203,14 @@
 
   (defconfig
     :app {sentry-url :sentry-dsn}
```
the hash-maps are no longer supported. Should this be updated to the vector version
[sentry-url :sentry-dsn]

briprowe replied:
Yes indeed

Commit:
Fix valid-app-config?

Commit:
Update defconfigs in test code.

Commit:
Fix bug in determining which entries are required

Commit:
Update code in clj-config.core's comment

Commit:
Fix valid-app-config?

briprowe commented:
Implements rentpath/clj-config#15

Commit:
Add the ability to pass a parse-fn function.

This adds the ability to specify an optional parse-fn
function in the config definition.

cc: @briprowe

briprowe commented:
Any objections to merging this?

@whalesalad @mayankag
